### PR TITLE
introduce consistent error handling

### PR DESCRIPTION
**Pull request by PulkitMishra**
@diptanu idea is to have an exceptions.py module which defines all custom exceptions, translation function for converting errors and a wrapper that can be added to extractors to better raise/handle errors. have set up a basic logger as well. the way these exception classes, the wrapper and logger will be used is explained with an example in moondream extractor and run_local cli command (needed to modify base_extractor a little as well). this more or less will be replicated across other extractors if it checks out for you. maybe the wrapper overcomlicates it a little and we could do without it as well. seemed neat to me

other modifications needed but not done in the draft pr
- tests
- docs

### CODE REVIEW

File: image/moondream/moondream_extractor.py [hunk 1]
```diff
@@ -1,65 +1,92 @@
...
+            self.tokenizer = AutoTokenizer.from_pretrained(model_id, revision=revision)
+        except Exception as e:
+            logger.error(f"Failed to initialize MoondreamExtractor: {str(e)}")
+            raise ModelLoadError(f"Failed to load Moondream model: {str(e)}")
 
-        if torch.cuda.is_available():
-            model = model.to("cuda")
+    @safe_execute
+    def extract(self, content: Content, params: MoondreamConfig) -> List[Union[Feature, Content]]:
+        if content.content_type not in self.input_mime_types:
```
Comment (diptanu) [inline image/moondream/moondream_extractor.py:52]: This is never going to happen. The scheduler will not send data that an extractor can't handle. Unless the user deliberately puts in the wrong mime type or we make some mistakes.

File: image/moondream/moondream_extractor.py [hunk 2]
```diff
@@ -1,65 +1,92 @@
...
+            answers = self.model.batch_answer([image], [prompt], self.tokenizer)
+            
+            results = []
+            for answer in answers:
+                results.append(Content.from_text(answer))
+            
+            return results
+        except Exception as e:
+            logger.error(f"Error during extraction: {str(e)}")
+            raise ExtractionProcessError(f"Failed to process image: {str(e)}")
```
Comment (diptanu) [inline image/moondream/moondream_extractor.py:68]: This is great. I think we need to track whether something is retrievable or not. Like an OpenAI call timing out should be retriable.

File: extractor-sdk/indexify_extractor_sdk/main.py [hunk 3]
```diff
@@ -67,7 +71,35 @@ def run_local(
...
+    try:
+        if text and file:
+            raise InputValidationError("You can only pass either text or file, not both.")
+        if not text and not file:
+            raise InputValidationError("You need to pass either text or file")
+
+        indexify_extractor.local(extractor, text, file)
+
+    except InputValidationError as e:
+        logger.error(f"Input Validation Error: {str(e)}")
```
Comment (diptanu) [inline extractor-sdk/indexify_extractor_sdk/main.py:83]: Is there any value of logging and also passing it along to typer to point on stdout?

File: extractor-sdk/indexify_extractor_sdk/main.py [hunk 4]
```diff
@@ -67,7 +71,35 @@ def run_local(
...
+    except IndexifyError as e:
+        logger.error(f"Indexify Error: {str(e)}")
+        typer.echo(f"Indexify Error: {str(e)}", err=True)
+        raise typer.Exit(code=3)
+    except Exception as e:
+        logger.exception("Unexpected error occurred")
+        typer.echo(f"An unexpected error occurred: {str(e)}", err=True)
+        raise typer.Exit(code=4)
+
+def get_mime_type(file_path):
```
Comment (diptanu) [inline extractor-sdk/indexify_extractor_sdk/main.py:99]: Where is this being used? Couldn't find any references

File: extractor-sdk/indexify_extractor_sdk/logging_config.py [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import logging
+from logging.handlers import RotatingFileHandler
+
+def configure_logging(level=logging.INFO):
```
Comment (diptanu) [inline extractor-sdk/indexify_extractor_sdk/logging_config.py:4]: Do we need this? Usually you just want to write to stdout and stderr, and let sysadmins figure out log rotation, log shipping and all that. Every company does their own thing.

File: extractor-sdk/indexify_extractor_sdk/exceptions.py [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+import torch
+from transformers import TensorflowAutoModelException, PyTorchAutoModelException
+from openai import OpenAIError
+from google.api_core import exceptions as google_exceptions
+from azure.core import exceptions as azure_exceptions
+from botocore import exceptions as aws_exceptions
+from pydantic import ValidationError
+from PIL import UnidentifiedImageError
+
+class IndexifyError(Exception):
```
Comment (diptanu) [inline extractor-sdk/indexify_extractor_sdk/exceptions.py:15]: Can we get a serialized version of the error - Like something that can be serialized to json? We could add an API on the server where it accepts the serialized form of exception so that users can see what happened on the data plane without looking at logs. This is a very common complain we get from users.

File: extractor-sdk/indexify_extractor_sdk/base_extractor.py [hunk 7]
```diff
@@ -258,11 +260,48 @@ def sample_html(self, features: List[Feature] = []) -> Content:
...
+        """
+        Validate that the input content type is supported by this extractor.
+        """
+        if content.content_type not in self.input_mime_types:
+            logger.error(f"Unsupported content type {content.content_type} for extractor {self.name}")
+            return False
+        return True
+
+
+def load_extractor(name: str) -> Type[Extractor]:
```
Comment (diptanu) [inline extractor-sdk/indexify_extractor_sdk/base_extractor.py:291]: Would appreciate if you can add some tests for all this! We developed the product so quickly that the python stuff has been very neglected and none of our tests on the data plane are in a good place. Would love to change direction here and have a healthy test suite for all these core features.

**Review by diptanu — CHANGES_REQUESTED** [review]
Looking great @PulkitMishra, I left some comments and appreciate your contribution. Are the methods in the base_extractor being used?

**Comment by PulkitMishra** [comment]
thanks for the review. gives me a good idea of what you are looking for. will work on it this weekend

File: extractor-sdk/indexify_extractor_sdk/base_extractor.py (hunk 7 above)
Comment (PulkitMishra) [reply extractor-sdk/indexify_extractor_sdk/base_extractor.py:274]: yeah i get it. i think its only fair to sacrifice on code quality a bit in lieu of speed in the beginning. i have written some tests for the code i have written and will push in a while but in general as you also mention i dont quite like how the entire test suite is as of now. for example there is a indexify-extractors/extractor-sdk/indexify_extractor_sdk/test/test_agent.py and also a indexify-extractors/extractor-sdk/tests/test_agent.py which is kind of weird. would like to change that. should i do it in this PR itself or treat it as a separate issue?

File: extractor-sdk/indexify_extractor_sdk/exceptions.py (hunk 6 above)
Comment (PulkitMishra) [reply extractor-sdk/indexify_extractor_sdk/exceptions.py:15]: added a to_json method in IndexifyError base class for this

File: extractor-sdk/indexify_extractor_sdk/logging_config.py (hunk 5 above)
Comment (PulkitMishra) [reply extractor-sdk/indexify_extractor_sdk/logging_config.py:4]: think its a tradeoff between simplicity vs control and where do you want the control to be? something i have done in the past for example is application -> log file -> fluent -> DB (via kafka) OR observabilty dashboards but that serves for a different usecase.
what we can maybe do is make it configurable but let the server define the configurations? (and ofc have some defaults)

File: extractor-sdk/indexify_extractor_sdk/main.py (hunk 4 above)
Comment (PulkitMishra) [reply extractor-sdk/indexify_extractor_sdk/main.py]: yeah not being used. original intention was to determine the MIME type of a file based on its extension to be used when dealing with file inputs where the MIME type isn't explicitly provided for cli. removing for now

File: extractor-sdk/indexify_extractor_sdk/main.py (hunk 3 above)
Comment (PulkitMishra) [reply extractor-sdk/indexify_extractor_sdk/main.py:83]: two reasons
1. both are for different audiences and usecases - logging provides a persistent record for debugging and analysis for developers, while typer output offers immediate feedback to the user of the CLI
2. logging lets you have different levels of detail. logs can contain more detailed information useful for debugging, while typer output can be more concise and user-friendly.

File: image/moondream/moondream_extractor.py (hunk 2 above)
Comment (PulkitMishra) [reply image/moondream/moondream_extractor.py]: introduced an additional retriable attribute in ExtractionProcessError class for this.

File: image/moondream/moondream_extractor.py (hunk 1 above)
Comment (PulkitMishra) [reply image/moondream/moondream_extractor.py]: yeah if the scheduler will never send wrong data to extractors then this is redundant and can be removed (and the validate_input method in base extractor as well).

**Comment by PulkitMishra** [comment]
@diptanu

1. handle_initialization_error : the idea is to handle errors that occur during the initialization of an extractor. it's not being used in the current implementation but the purpose would be to provide a standardized way to handle and log errors that occur when creating an instance of an extractor. these initialization errors could be handled directly in the __init__ method of each extractor as well, so if we want we can remove it from here.
2. validate_input : this was to check if the content type of the input is supported by the extractor. however, as pointed out by you this check is redundant because the scheduler will not send data that an extractor can't handle. i kind of missed that. Given this, we can now consider removing this method as well
3. all comments other than logging and testing have been addressed. would like your opinions on the same

File: extractor-sdk/indexify_extractor_sdk/main.py (hunk 3 above)
Comment (diptanu) [reply extractor-sdk/indexify_extractor_sdk/main.py:83]: Makes sense. I agree with you.

File: extractor-sdk/indexify_extractor_sdk/logging_config.py (hunk 5 above)
Comment (diptanu) [reply extractor-sdk/indexify_extractor_sdk/logging_config.py:4]: @PulkitMishra Yeah it's perfectly fine to define logging levels, I think we can skip configuring log rotation at this layer.

File: extractor-sdk/indexify_extractor_sdk/exceptions.py (hunk 6 above)
Comment (diptanu) [reply extractor-sdk/indexify_extractor_sdk/exceptions.py:15]: Love it, amazing.

File: extractor-sdk/indexify_extractor_sdk/base_extractor.py (hunk 7 above)
Comment (diptanu) [reply extractor-sdk/indexify_extractor_sdk/base_extractor.py:274]: Yeah let's do it in a separate issue. We should do this as a team and revamp how testing is done for the data plane.

**Comment by diptanu** [comment]
@PulkitMishra this is great, I don't have any other comments. The direction we are headed towards is - 
 1. Move extractor related abstractions into the `indexify` python client. 
 2. Make the data plane be responsible for downloading serialized code or code the way it downloads now and caches, runs it.
 3. We have the first version of this checked in with the client library. 
 
 Let's merge your PR and then I can move some of the exception handling code connected with the extractors into the client library or you could move them yourself.

**Comment by diptanu** [comment]
@PulkitMishra Any update on this? I have changed the data plane significantly and things should be much cleaner now! Let's move this to finish line.
